ddtrace/internal/processor/stats.py

```diff
@@ -0,0 +1,239 @@
+            bucket_time_ns = span_end_ns - (span_end_ns % self._bucket_size_ns)
+            aggr_key = _span_aggr_key(span)
+            stats = self._buckets[bucket_time_ns][aggr_key]
+
+            stats.hits += 1
+            stats.duration += span.duration_ns
+            if is_top_level:
+                stats.top_level_hits += 1
+            if span.error:
+                stats.errors += 1
+                stats.err_distribution.add(span.duration_ns)
+            else:
+                stats.ok_distribution.add(span.duration_ns)
+
+    @property
+    def _agent_endpoint(self):
+        return "%s%s" % (self._agent_url, self._endpoint)
+
+    def _serialize_buckets(self):
+        # type: () -> List
+        """Serialize and update the buckets.
+
+        The current bucket is left in case any other spans are added.
+        """
+        serialized_buckets = []
+        serialized_bucket_keys = []
+        now_ns = time_ns()
+        for bucket_time_ns, bucket in self._buckets.items():
+            # TODO: have to be able to force these out on process shutdown
+            if bucket_time_ns > now_ns - self._bucket_size_ns:
+                # do not flush the current bucket
+                # DEV: is this actually required? Can a bucket not be reported twice?
+                #      What about spans `finished` with a custom end time (in the past)
+                #      after the bucket has been flushed?
+                continue
+            bucket_aggr_stats = []
+            serialized_bucket_keys.append(bucket_time_ns)
+
+            for aggr_key, stat_aggr in bucket.items():
+                name, service, resource, _type, http_status, synthetics = aggr_key
+                bucket = {
+                    "Name": name,
+                    "Resource": resource,
+                    "Synthetics": synthetics,
+                    "Hits": stat_aggr.hits,
+                    "TopLevelHits": stat_aggr.top_level_hits,
+                    "Duration": stat_aggr.duration,
+                    "Errors": stat_aggr.errors,
+                    "OkSummary": DDSketchProto.to_proto(stat_aggr.ok_distribution).SerializeToString(),
+                    "ErrorSummary": DDSketchProto.to_proto(stat_aggr.err_distribution).SerializeToString(),
+                }
+                if service:
+                    bucket["Service"] = service
+                if _type:
+                    bucket["Type"] = _type
+                bucket_aggr_stats.append(bucket)
+            serialized_buckets.append(
+                {
+                    "Start": bucket_time_ns,
+                    "Duration": self._bucket_size_ns,
+                    "Stats": bucket_aggr_stats,
+                }
+            )
+
+        # Clear out buckets that have been serialized
+        for key in serialized_bucket_keys:
+            del self._buckets[key]
+
+        return serialized_buckets
+
+    def periodic(self):
+        # type: (...) -> None
+        with self._lock:
```

Comment:
we can consider moving this lock into `_serialize_buckets`

Reply:
Hmm I think that would be misleading. The reason it is required it because it's called in `periodic` so I think the lock makes sense here.

Reply:
this makes sense, I only worry about someone accidentally calling `_serialize_buckets` outside of `periodic` and then having issues.

We could move to an `RLock` and add locking to both, but probably just unnecessary overhead. Maybe we can add a comment/docstring to `_serialize_buckets` saying that it can only be called when `self._lock` is acquired? 🤷🏻‍♂️

Reply:
Outside of periodic is fine though because then it would be called from the main thread and not have any race condition.